lib/sshutils/sftp/parse.go

```diff
@@ -17,64 +17,103 @@ limitations under the License.
 )
 
-var reSFTP = regexp.MustCompile(
-	// optional username, note that outside group
-	// is a non-capturing as it includes @ signs we don't want
-	`(?:(?P<username>.+)@)?` +
-		// either some stuff in brackets - [ipv6]
-		// or some stuff without brackets and colons
-		`(?P<host>` +
-		// this says: [stuff in brackets that is not brackets] - loose definition of the IP address
-		`(?:\[[^@\[\]]+\])` +
-		// or
-		`|` +
-		// some stuff without brackets or colons to make sure the OR condition
-		// is not ambiguous
-		`(?:[^@\[\:\]]+)` +
-		`)` +
-		// after colon, there is a path that could consist technically of
-		// any char including empty which stands for the implicit home directory
-		`:(?P<path>.*)`,
-)
-
-// Destination is SCP destination to copy to or from
+// Destination is a remote SFTP destination to copy to or from.
 type Destination struct {
 	// Login is an optional login username
 	Login string
 	// Host is a host to copy to/from
-	Host utils.NetAddr
+	Host *utils.NetAddr
 	// Path is a path to copy to/from.
 	// An empty path name is valid, and it refers to the user's default directory (usually
 	// the user's home directory).
 	// See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 	Path string
 }
 
-// ParseSCPDestination takes a string representing a remote resource for SFTP
-// to download/upload, like "user@host:/path/to/resource.txt" and parses it into
+// ParseDestination takes a string representing a remote resource for SFTP
+// to download/upload in the form "[user@]host:[path]" and parses it into
 // a structured form.
 //
 // See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 // section about details on file names.
-func ParseDestination(s string) (*Destination, error) {
-	out := reSFTP.FindStringSubmatch(s)
-	if len(out) < 4 {
-		return nil, trace.BadParameter("failed to parse %q, try form user@host:/path", s)
+func ParseDestination(input string) (*Destination, error) {
+	firstColonIdx := strings.Index(input, ":")
+	if firstColonIdx == -1 {
+		return nil, trace.BadParameter("%q is missing a path, use form [user@]host:[path]", input)
+	}
+	loginEndIdx := strings.LastIndex(input[:firstColonIdx], "@")
+	// if a login exists and the path begins right after the login ends,
+	// no host is specified
+	if loginEndIdx != -1 && loginEndIdx+1 == firstColonIdx {
+		return nil, trace.BadParameter("%q is missing a host, use form [user@]host:[path]", input)
 	}
-	addr, err := utils.ParseAddr(out[2])
+
+	var login string
+	// If at least one '@' exists and is before the first ':', get the
+	// login. Otherwise, either there are no '@' or all '@' are after
+	// the first ':' (where the host or path starts), so no login was
+	// specified.
+	if loginEndIdx != -1 {
+		login = input[:loginEndIdx]
+		loginEndIdx++
+	} else {
+		loginEndIdx = 0
+	}
+
+	hostEnd := firstColonIdx + 1
+	host, err := utils.ParseAddr(input[loginEndIdx:firstColonIdx])
 	if err != nil {
```

Comment:
Instead of relying on `ParseAddr` to return an error, wouldn't it make more sense to first check whether the first character after `@` (`loginEndIdx`) is `[`?

- If it is, then look for `]`, parse the address (or return error if there's no closing square bracket).
  - Then make sure next character is `:` and take everything after it as path.
- If it's not, parse everything until the first `:` as address, and then take everything after it as path.

Parsing logic flow would be a bit more straightforward this way, no?

Reply:
Good find, I was thinking someone might uncover a simpler way to write this when reviewing, you didn't disappoint :)